clang/lib/Sema/SemaBounds.cpp

```diff
@@ -1896,15 +2004,53 @@ namespace {
      Cfg->print(llvm::outs(), S.getLangOpts(), true);
      llvm::outs() << "Traversing CFG:\n";
 #endif
+
+     // Map each function parameter to its declared bounds (if any) before
+     // checking the body of the function.  The context formed by the declared
+     // parameter bounds is the initial context for checking the function body.
+     CheckingState ParamsState;
+     for (auto I = FD->param_begin(); I != FD->param_end(); ++I) {
+       ParmVarDecl *Param = *I;
+       BoundsExpr *Bounds = Param->getBoundsExpr();
+       if (Bounds)
+         ParamsState.UC[Param] = Bounds;
+     }
+     llvm::DenseMap<unsigned int, CheckingState> BlockStates;
+     BlockStates[Cfg->getEntry().getBlockID()] = ParamsState;
+
      StmtSet NestedElements;
      FindNestedElements(NestedElements);
      StmtSet MemoryCheckedStmts;
      StmtSet BoundsCheckedStmts;
      IdentifyChecked(Body, MemoryCheckedStmts, BoundsCheckedStmts, CheckedScopeSpecifier::CSS_Unchecked);
+
+     // Run the bounds widening algorithm on this function.
+     BoundsAnalysis BA = getBoundsAnalyzer();
+     BA.WidenBounds(FD);
+     if (S.getLangOpts().DumpWidenedBounds)
+       BA.DumpWidenedBounds(FD);
+
      PostOrderCFGView POView = PostOrderCFGView(Cfg);
      ResetFacts();
      for (const CFGBlock *Block : POView) {
        AFA.GetFacts(Facts);
+       CheckingState BlockState = GetIncomingBlockState(Block, BlockStates);
+
+       // Get the widened bounds for the current block as computed by the
+       // bounds widening algorithm.
+       BoundsMapTy WidenedBounds = BA.GetWidenedBounds(Block);
+       // Also get the bounds killed (if any) by each statement in the current
+       // block.
+       StmtDeclSetTy KilledBounds = BA.GetKillSet(Block);
+
+       // Bounds killed by a statement in the current block remain killed
+       // through the end of the block. So we remove those bounds from the
+       // widened bounds for the current block.
+       for (const auto item : KilledBounds) {
+         for (const VarDecl *V : item.second)
+           WidenedBounds.erase(V);
```

Comment:
Does this kill bounds widened in the current block even for statements that occur before the bounds are killed? E.g.
```
void f(nt_array_ptr p : count(len), int len) {
    if (*p) { // widen bounds of p
        p[1]; // bounds of p should still be widened here
        len = 0; // kill the widened bounds of p
    }
}
```
It seems like the widened bounds of `p` will be killed for the entire block, before the statement `len = 0;` kills them.

Reply:
Yes, my original intent was to remove from WidenedBounds only the bounds killed by the current statement. But the problem is that BlockState.UC would also need to be updated to set the bounds for the VarDecl to the declared bounds.
 For this GetDeclaredBounds would need to be called outside the loop and the context would need to be stored in a declared context.
Otherwise, this case does not work:
`
nt_array_ptr p : count(0);
int i;
if (*p) {
  i = 0;
  p++;
  if (*(p + 1)) {}
}
Here, the BlockState.UC is updated with p:1 inside the block `if (*p)`. Then the bounds are killed at `p++`. But the BlockState.UC still has `p:1`. So in this case we also need to reset the context with the declared bounds for p which are `p:0`.